### ISSUE TITLE
Error: EHOSTUNREACH prevents connecting to bond

### ISSUE THREAD

**Issue by altrahm** [aarons22/homebridge-bond#53]
Here's the debug log:
```[16/11/2019, 09:04:29] [Bond] DEBUG: Request [GET http://192.168.0.16/v2/devices] - body: {}
[16/11/2019, 09:04:29] Homebridge is running on port 51826.
[16/11/2019, 09:04:32] [Bond] DEBUG: Error [GET http://192.168.0.16/v2/devices] - {"name":"RequestError","message":"Error: connect EHOSTUNREACH 192.168.0.16:80","cause":{"errno":-113,"code":"EHOSTUNREACH","syscall":"connect","address":"192.168.0.16","port":80},"error":{"errno":-113,"code":"EHOSTUNREACH","syscall":"connect","address":"192.168.0.16","port":80},"options":{"method":"GET","uri":"http://192.168.0.16/v2/devices","headers":{"BOND-Token":"ef762345c434fbf1"},"body":{},"json":true,"simple":false,"timeout":10000,"resolveWithFullResponse":false,"transform2xxOnly":false}}
[16/11/2019, 09:04:32] [Bond] ERR: A request error occurred: Error: connect EHOSTUNREACH 192.168.0.16:80
[16/11/2019, 09:04:32] [Bond] Error getting device ids: TypeError: Cannot convert undefined or null to object
[16/11/2019, 09:04:32] [Bond] 0 cached accessories were loaded
[16/11/2019, 09:04:32] [Bond] Getting devices...
[16/11/2019, 09:04:32] [Bond] 0 devices were found on this Bond.
[16/11/2019, 09:04:32] [Bond] Attempting to add 0 devices that were not previously added.
```

**Information:**
 - Plugin version: homebridge-bond@2.5.0
 - Bond firmware version: v2.9.5

**Additional context**
I'm using the bond's local ip in the config file. Using this worked a couple of times, but would stop after a little (not sure how long, maybe 10 minutes?). I tried clearing the cachedAccessories and persist folders and re-adding the device to the Home app. I'm fairly certain it's the right ip, since going to it in the browser brings up a bond screen where I can view the network it's connected to.
I also tried using the ip obtained by pinging the bond ip. That one never worked, but it doesn't give this error (it still says `Error getting device ids: TypeError: Cannot convert undefined or null to object`)

**Comment by aarons22** [comment]
It sounds like the bond is unable to resolve your IP address. Can you confirm that running the curl command found in the bond v2 api docs works for you? http://docs-local.appbond.com/#section/Getting-Started/Check-Bond-Version

```
curl -i http:/<BOND_IP>/v2/sys/version
```

**Comment by aarons22** [comment]
@altrahm happy to help with this issue but I need more info. If not, I'll go ahead and close the issue.

**Comment by thomasmaxson** [comment]
I am having this exact issue. I followed the directions, and can confirm that I am receiving a 200 OK header response code.

**Comment by rhuss2** [comment]
I’m getting this same error. 
I have 4 fans setup on Bond. 

Homebridge] [1/20/2020, 5:27:59 PM] [Bond] ERR: A request error occurred: Error: getaddrinfo ENOTFOUND undefined undefined:80
[Homebridge] [1/20/2020, 5:27:59 PM] [Bond] Error getting device ids: TypeError: Cannot convert undefined or null to object
[Homebridge] [1/20/2020, 5:27:59 PM] [LG OLED 65 in TV] webOS - connected to TV
[Homebridge] [1/20/2020, 5:27:59 PM] [Bond] 0 cached accessories were loaded
[Homebridge] [1/20/2020, 5:27:59 PM] [Bond] Getting devices...
[Homebridge] [1/20/2020, 5:27:59 PM] [Bond] 0 devices were found on this Bond.
[Homebridge] [1/20/2020, 5:27:59 PM] [Bond] Attempting to add 0 devices that were not previously added.

**Comment by thomasmaxson** [comment]
@aarons22 is there any help/assistance that you can offer with these issues?

**Comment by aarons22** [comment]
A couple questions/thoughts:
- Is your homebridge being run through a docker container or something of the like? 
- Can you run this plugin in debug mode and provide the logs? You can find instructions for that [here](https://github.com/aarons22/homebridge-bond#debug-mode)
- Did your bond account get migrated from v1? The error `Cannot convert undefined or null to object` leads me to believe the data from your devices is not formatted correctly. I know it's not ideal, but it might be good to try clearing your bond devices and re-adding through their app.
- Lastly, if you are up for it, you can always try messing around with the code yourself locally to see if you can debug the issue on your end: https://github.com/aarons22/homebridge-bond#development

**Comment by cgapperi** [comment]
Just running into this issue myself. Things were running just dandy then, Spectrum reset my router for some reason. IP address for Bond was changed, but updated config file. Then I noticed an update to Homebridge. Did that and also a firmware update to Bond. Not sure which has caused the issue. Anyone else having similar issues with 

Bond firmware: v2.10.21
Homebridge: homebridge-config-ui-x v4.13.3
Homebridge-bond: homebridge-bond v2.5.4

I suspect it has to do with the firmware update because Bond is no longer reachable from any device on the network, but it functions just fine in Bond v2 iOS app

**Comment by chrismerck** [comment]
If the Bond is operable from the Bond Home app, but it is not reachable on the local network, it is likely that the IP address you are using to reach the Bond is no longer correct. The IP could have been reassigned by your network's DHCP server. I suspect that something in mDNS (not sure where the fault lies) could be caching old values of the IP address.

If someone catches this issue live, you could try comparing the IP that HomeBridge is trying to reach with the IP in your DHCP client table (probably accessible via your WiFi router's admin page).

---

@aarons22  Two questions:  (1) does HomeBridge support listening on user-specified UDP ports? Specifically 30007? (This is disallowed on iOS and Android, but allowed on MacOS and Linux.) and  (2)  does HomeBridge support opening a UDP broadcast socket and then receiving replies from multiple 3-rd parties? (AFAIK this is allowed on every system supporting BSD sockets, however higher-level utilities (nc) and libraries often overlook this use-case.)

If either of the above are possible, we may be able to facilitate faster more reliable discovery and IP change detection using BPUP.

**Comment by aarons22** [comment]
@chrismerck Yup, I was able to get BPUP working with some basic commands when I tried it out a few week back. How would we be able to detect a change in IP?

**Comment by aarons22** [comment]
Closing this issue as it's pretty old and I'm hoping one of the releases since has resolved the issue (specifically, 3.0.0, which uses BPUP).
